File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/messaging/EventBus.java
package pe.joedayz.microservicios.modulo01.messaging;

import pe.joedayz.microservicios.modulo01.ddd.shared.DomainEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

/**
 * Bus de eventos MUY simplificado que simula Apache Kafka (Modulo 4).
 *
 * <p>Los productores publican eventos y los consumidores suscritos los reciben. En el
 * curso real esto es un topic de Kafka con consumer groups; aqui es sincrono y en memoria
 * para que el ejemplo sea autocontenido.
 */
public class EventBus {

    private final List<Consumer<DomainEvent>> subscribers = new ArrayList<>();

    public void subscribe(Consumer<DomainEvent> subscriber) {
        subscribers.add(subscriber);
    }

    public void publish(DomainEvent event) {
        System.out.printf("   [Kafka] publicando %s (tenant=%s)%n", event.eventType(), event.tenantId());
        for (Consumer<DomainEvent> subscriber : subscribers) {
            subscriber.accept(event);
        }
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Messaging/EventBus.cs
using JoeDayz.Microservicios.Modulo01.Ddd.Shared;

namespace JoeDayz.Microservicios.Modulo01.Messaging;

/// <summary>
/// Broker simulado (en produccion: Kafka, Azure Service Bus o Amazon SNS/SQS).
/// Lo importante didacticamente es el desacople: quien publica no conoce a quien consume.
/// </summary>
public sealed class EventBus
{
    private readonly List<Action<IDomainEvent>> _subscribers = [];

    public void Subscribe(Action<IDomainEvent> subscriber) => _subscribers.Add(subscriber);

    public void Publish(IDomainEvent domainEvent)
    {
        Console.WriteLine($"   [Kafka] publicando {domainEvent.EventType} (tenant={domainEvent.TenantId})");
        foreach (var subscriber in _subscribers)
        {
            subscriber(domainEvent);
        }
    }
}
